### ISSUE TITLE
google_sql_database_instance.settings.backup_configuration.binary_log_enabled is not changing

### ISSUE THREAD

**Issue by syedrakib** [terraform-providers/terraform-provider-google#6121]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

0.12.24

### Affected Resource(s)

* google_sql_database_instance

### Synopsis

I spinned up ClousSQLMysQL with `settings.backup_configuration.binary_log_enabled` set to `true`. Now i am trying to set it to `false` with this

            backup_configuration = {
              enabled            = false
              binary_log_enabled = false
              start_time         = "00:05"
            }

---

The `terraform plan` said:

            backup_configuration {
              ~ binary_log_enabled = true -> false
                enabled            = false
                start_time         = "00:05"
            }

---

The `terraform apply` was successful saying

    Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

### The Issue

However, when i plan and apply again, it is again saying the same things as if the previous changes were not performed.

### Verification

I also looked into the TFState file and it shows the following even though the `plan-apply` steps said it changed `backup_configuration.binary_log_enabled` to `false` successfully

    "backup_configuration": [
        {
            "binary_log_enabled": true,
            "enabled": false,
            "location": "",
            "start_time": "00:05"
        }
    ],

**Comment by syedrakib** [comment]
When i tried changing **"Enable point-in-time recovery"** (of CloudSQL MySQL) from `true` to `false` using **the GCP console**, it showed a confirmation pop-up that CloudSQL requires a restart for the changes to take effect. I don't think we are invoking that restart when we are change `settings.backup_configuration.binary_log_enabled` from `true` to `false` via Terraform - the "modification" state completes and returns to "modification complete" almost immediately.

**Comment by syedrakib** [comment]
When we change `settings.backup_configuration.binary_log_enabled` from `true` to `false` using **Terraform** and then we do `terraform refresh -target=resource`, the state of this resource is not actually refreshed - the refreshed value for `settings.backup_configuration.binary_log_enabled` still stays at `true`

**Comment by venkykuberan** [comment]
@syedrakib if backup_configuration block is `enabled` then API sets the binary_log_enabled to `true` although you set it to false in your config. Terraform updates the state file based on the response from API. If you want binary_log_enabled to be false don't use that block in your config API sets it to false.

doc -  https://www.terraform.io/docs/providers/google/r/sql_database_instance.html

binary_log_enabled - (Optional) True if binary logging is enabled. If settings.backup_configuration.enabled is false, this must be as well. 

Let me know if i can close this issue if it fix your problem

**Comment by syedrakib** [comment]
I don't quite understand what is meant by "API sets the `binary_log_enabled` to true although you set it to false". I mean, defeats the purpose of having a `binary_log_enabled` argument then?

:/

According to the referred doc, if `settings.backup_configuration.enabled` is `false` then this `settings.backup_configuration. binary_log_enabled ` must also be false.

Here is what i have in my backup_configuration:

    backup_configuration = {
      enabled            = false
      binary_log_enabled = false
      start_time         = "00:05"
    }

That complies with the doc, right? The expectation here is, backup is disabled, binary_log is disabled. Thus, it should set it `false` and the API response should return `false` instead of `true`.

**Comment by syedrakib** [comment]
According to the referred doc, if `settings.backup_configuration.enabled` is `false` then this `settings.backup_configuration. binary_log_enabled ` must also be false.

Here is what i have in my backup_configuration:

    backup_configuration = {
      enabled            = false
      binary_log_enabled = false
      start_time         = "00:05"
    }

That complies with the doc, right? The expectation here is, backup is disabled, binary_log is disabled. Thus, it should set it `false` and the API response should return `false` instead of `true`.

**Comment by venkykuberan** [comment]
When backup is disabled, binary_log should also be disabled, API accepts it.  If backup is disabled and binary_log is enable API throws

```Error: Error, failed to create instance test-sql6: googleapi: Error 400: Invalid request: Binary log must be disabled when backup is disabled.., invalid ```

 The following state is invalid, i am not sure how apply command worked for you. 
``
"backup_configuration": [
    {
        "binary_log_enabled": true,
        "enabled": false,
        "location": "",
        "start_time": "00:05"
    }
]
```
can you share your provider version ?

**Comment by syedrakib** [comment]
Hi @venkykuberan . Thanks for the reply. Yes that's exactly the bug i guess how come it worked on the first place.

I am using these provider versions:

```
terraform {
  required_version = "0.12.24"
}

provider "google" {
  version = "3.13.0"
}

provider "google-beta" {
  version = "3.13.0"
}
```

**Comment by syedrakib** [comment]
Just now i have manually changed the value of `backup_configuration . binary_log_enabled` from `true -> false` inside the TFState file.

My main.tf file still says `backup_configuration . binary_log_enabled = false`

Then i tried `terraform plan` & `terraform apply` and the value of `backup_configuration . binary_log_enabled` inside the TFState file changed to `true` again.

I check the CloudSQL in GCP. Both the values for **Automated Backups** & **Enable point-in-time recovery** are both set to unchecked (false).

Why is terraform seeing it `true` then?

**Comment by venkykuberan** [comment]
`Just now i have manually changed the value of backup_configuration . binary_log_enabled from true -> false inside the TFState file. `
  Terraform state file shouldn't edited manually. The flow is 

- Change the desired state in config file (.tf)
-  run `terraform plan` => checks syntactical errors and run other internal funcs (No API call is made yet)
- run `terraform apply` => API call is made and the request is success, terraform state file is updated with the API response.

Please follow the above steps, for what ever changes you want to make in config and verify your change is accepted.

**Comment by syedrakib** [comment]
Ok. Ignore that. That was me just sharing that even when i explicitly change the value in the sate file myself, it reverts back to what it was.

**Comment by syedrakib** [comment]
The bottom-line of the problem is.....

in GCP CloudSQL MySQL instance:
* the auto-backup feature is checked off
* the point in time recover featue is checked off
* which implies the binary_log feature is also turned off

Then why is it that the `google_sql_database_instance` resource "refreshing" and setting the value of `settings.backup_configuration.binary_log_enabled` to `true`?

**Comment by syedrakib** [comment]
Hi... just following up on the issue with this one..... could we identify the bug with this?

**Comment by chrisst** [comment]
Fix will be going out in 3.19.0
